### PR TITLE
Use _get_distribution_strategy only when it is available.

### PR DESCRIPTION
**Pull request by hgaiser**
The changes introduced in https://github.com/tensorflow/tensorflow/commit/06d8f77bcc43826eb2b30209a73f3c833079edbc are not compatible with standalone `Keras` (they are compatible with `tf.Keras`). a `keras.Model` does not have a `_get_distribution_strategy` method, which is now assumed for the `Tensorboard` callback.

These kind of differences are usually not an issue, but Keras is using [`tf.keras.callbacks.Tensorboard`](https://github.com/keras-team/keras/blob/master/keras/callbacks/tensorboard_v2.py#L18) instead of implementing its own.

This PR checks if `_get_distribution_strategy` exists and uses it if it does, making it compatible with a `keras.Model`.

### CODE REVIEW

**Review by rchao — CHANGES_REQUESTED** [review]
Thanks for the PR! Would you mind adding some comment above the `if-else` so it's clear the rationale for it?

**Comment by hgaiser** [comment]
Done :+1:

File: tensorflow/python/keras/callbacks.py [hunk 1]
```diff
@@ -1526,10 +1526,14 @@ def set_model(self, model):
     """Sets Keras model and writes graph if specified."""
     self.model = model
 
-    # TensorBoard callback involves writing a summary file in a
-    # possibly distributed settings.
-    self._log_write_dir = distributed_file_utils.write_dirpath(
-        self.log_dir, self.model._get_distribution_strategy())  # pylint: disable=protected-access
+    # In case this callback is used via Keras, _get_distribution_strategy does not exist.
```
Comment (rchao) [inline tensorflow/python/keras/callbacks.py]: I would disambiguate it by saying keras-team Keras or native Keras

**Comment by gbaned** [comment]
@hgaiser Can you please check reviewer comments and keep us posted. Thanks!

**Comment by hgaiser** [comment]
Apologies, I forgot to update the PR. It is updated now though :+1:

**Comment by gbaned** [comment]
@rchao Can you please take a look on this PR? Thanks!

**Review by rchao — APPROVED** [review]
Change looks good. Thanks for the PR!

**Comment by aguldbrandsen** [comment]
I cant get Tensorboard to work when using [https://github.com/fizyr/keras-retinanet](keras-retinanet). @hgaiser is it fixed with this commit?

**Comment by hgaiser** [comment]
> I cant get Tensorboard to work when using [https://github.com/fizyr/keras-retinanet](keras-retinanet). @hgaiser is it fixed with this commit?

Should be, yes.

**Comment by aguldbrandsen** [comment]
> > I cant get Tensorboard to work when using [https://github.com/fizyr/keras-retinanet](keras-retinanet). @hgaiser is it fixed with this commit?
> 
> Should be, yes.

I am experiencing the same error as here: https://github.com/fizyr/keras-retinanet/issues/1239
Should have the latest version installed. 

Just to get it right: Is the commit not merged yet? And is that why the error is still occurring?

**Comment by hgaiser** [comment]
It's not merged yet, so unless you applied this branch yourself you're not using this fix.

**Comment by aguldbrandsen** [comment]
> It's not merged yet, so unless you applied this branch yourself you're not using this fix.

It works - Thank you!

**Comment by bwanaaa** [comment]
Still happening. I am using a pip installed version.
`!pip freeze | grep -iE "tensorflow|keras"`
gives
```
Keras==2.3.1
Keras-Applications==1.0.8
Keras-Preprocessing==1.1.0
tensorflow==2.1.0
tensorflow-estimator==2.1.0

```

Am I not using the latest??

```
$ pip search tensorflow
tensorflow (2.1.0)                                - TensorFlow is an open source machine learning framework for everyone.
  INSTALLED: 2.1.0 (latest)
tensorflow-qndex (0.0.22)                         - tensorflow-qnd x tensorflow-extenteten

```

**Comment by aguldbrandsen** [comment]
A Tensorflow update is missing for tid to work, so you need to go edit the callbacks.py file with the edit of this commit to get it to work at the moment. In your local Tensorflow installation just replace the entire callbacks.py file.

**Comment by bwanaaa** [comment]
thank you for your reply. I download tensorflow-master from here:

`https://github.com/tensorflow/tensorflow`

and I replaced the callbacks.py file but now get this error:

`ImportError: cannot import name 'profiler_v2' from 'tensorflow.python.profiler' (/home/stefan/miniconda3/envs/TFg/lib/python3.7/site-packages/tensorflow_core/python/profiler/__init__.py)
`

**Comment by bwanaaa** [comment]
I found profiler v2 and copied it over. Now I get this error

`ImportError: cannot import name '_pywrap_profiler' from 'tensorflow.python.profiler.internal' (/home/stefan/miniconda3/envs/TFg/lib/python3.7/site-packages/tensorflow_core/python/profiler/internal/__init__.py)
`
There is an __init.py__ file inside 'internal'

**Comment by bwanaaa** [comment]
Now, when I get to this step

```
#train for 50 epochs
autoencoder.fit(x_train, x_train,
                epochs=50,
                batch_size=256,
                shuffle=True,
                validation_data=(x_test, x_test),
                verbose=1)
```

I get this error

`ImportError: cannot import name '_pywrap_profiler' from 'tensorflow.python.profiler.internal' (/home/stefan/miniconda3/envs/TFg/lib/python3.7/site-packages/tensorflow_core/python/profiler/internal/__init__.py)

notice there are no callbacks to tensorboard-this is just running the simple VAE
`

**Review by Yannick947 — COMMENTED** [review]
Looks good but is still not merged, ist it?

**Comment by hgaiser** [comment]
Nope not merged yet.

**Comment by lihude** [comment]
AttributeError: 'Model' object has no attribute 'run_eagerly'

**Comment by sangasangasanga** [comment]
AttributeError: 'ResourceVariable' object has no attribute '_distribute_strategy'

**Pull request merged**